pkg/operator/controller/route-status/route_status.go

```diff
@@ -0,0 +1,273 @@
+package routestatus
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+)
+
+// A brief overview of the operator's interactions with route status:
+// Though the openshift-router is mainly responsible for route object status, the operator plays a small, but
+// significant role in ensuring the route status is accurate. The openshift-router updates the route object's status
+// when it is admitted to an ingress controller. However, the openshift-router is unable to reliably update the route's
+// status when it stops managing the route. Here are the scenarios where the operator steps in:
+// #1 When the ingress controller, the corresponding router deployment, and its pods are deleted.
+//    - The operator knows when a router is deleted because it is the one responsible for deleting it. So it
+//      simply calls ClearRouteStatus to clear status of routes that openshift-router has admitted.
+//    - Handled by the ingress control loop.
+// #2 When the ingress controller sharding configuration (i.e., selectors) is changed.
+//    - When the selectors (routeSelector and namespaceSelector) are updated, the operator simply clears the status of
+//      any route that it is no longer selecting using the updated selectors.
+//    - We determine what routes are admitted by the current state of the selectors (just like the openshift-router).
+//    - Handled by the ingress control loop.
+// # 3 When the route's labels are changed.
+//    - When the route labels are updated, the operator will reconcile that specific route and clean up any admitted
+//      status that is now stale.
+//    - Handled by the route-status control loop.
+// # 4 When the namespace's labels are changed.
+//    - When the namespace labels are updated, the operator will reconcile all the routes in the namespace and clean up
+//      stale route statuses.
+//    - Handled by the route-status control loop.
+
+// ClearAllRoutesStatusForIngressController clears any route status that has been admitted by the ingress controller,
+// regardless if it is actually admitted or not. This function should be used for deletions of ingress controllers.
+func ClearAllRoutesStatusForIngressController(kclient client.Client, icName string) []error {
+	// List all routes.
+	errs := []error{}
+	start := time.Now()
+	routeList := &routev1.RouteList{}
+	routesCleared := 0
+	if err := kclient.List(context.TODO(), routeList); err != nil {
```

Comment:
Does this list all routes in the entire cluster? That seems like that could be a lot of items to loop over. 

It looks like `List()` can filter based on namespace, selector, or labels that the items have (and maybe there are other filtering options I haven't seen). Would any of those be applicable to cut down on the number of items that `List()` returns?

Reply:
Good question. Full disclosure, I didn't change this logic in this PR (not saying I wont' fix it now though), but I did write it. Use https://github.com/openshift/cluster-ingress-operator/pull/849/commits/e2c30c47ef5ccf467d7a6ad924386c5a506c0147 for your comparison since `route_status.go` shows as added/removed in the PR diff.

Potentially, we could read the IC's route selector and namespace selector (if exists) and filter by those. In theory, we should only have to clear routes that are current selected by the selectors, however, clearing all of the routes is a bit more fail-safe. If another part of this route status logic didn't clear a status when say the routeSelector was updated, doing all routes would cover our tracks.

I agree it's a performance hit, but since this only happens when you delete an IC, I'd argue it's not that big of  problem. Do you agree that deleting an IC should be pretty rare?

Reply:
Yeah, I agree that IC deletion is pretty rare, and making sure you definitely catch every route makes sense.